Projects/Server/Network/NetState/NetState.cs

```diff
@@ -68,6 +70,14 @@ public partial class NetState : IComparable<NetState>, IValueLinkListNode<NetSta
     private long[] _packetCounts;
     private string _disconnectReason = string.Empty;
 
+    /**
+     * If this field is not null, then a packet handler is currently
+     * running asynchronously and handling of other packets should be
+     * suspended.
+     */
+    internal Task _packetHandlerTask;
```

Comment:
I may be misunderstanding, but, if this is going to block the handling of future packets, what do we gain by doing packet handling asynchronously in the first place?

Reply:
This blocks handling packets of this client only. What you gain is that the server can continue handling packets from other clients.
Blocking packet handling in the current client is necessary because otherwise you'd be handling packets in random order, and that will cause major pains. All current code assumes that packet handling (within one client) is serialized.

Reply:
> This blocks handling packets of this client only.

I noticed it's contained to the actual `NetState` of a single connection.  I'm mostly just curious what the use case is, if you're willing to share.  It's worked the same way for 20 years and I haven't ever given it much thought.

so, if I understand this correctly, you want to "block" subsequent packets until the `Task` is done so you can maintain ordering?

Reply:
Simple example: you implement an AccountHandler that does a network request to verify a password. If you do it synchronously, the whole shard freezes until the auth server responds. With this PR, it's possible to do such a network request asynchronously, and let the shard keep running while that one client is authenticating.

> you want to "block" subsequent packets until the `Task` is done so you can maintain ordering?

Yes, that's what I just said. It's not something I "want". It's technically necessary due to how the rest of the code base is implemented.

Reply:
/shrug

thanks.

nice improvement, in my opinion, "technically necessary" is a stretch, again in my opinion - we await Kamron re: CLA

Reply:
So, in your opinion, the current code base can handle out-of-order handling of packets?
This isn't something one can have an opinion on - it is either true or false. There's no room for opinions and no room for "stretching".

Reply:
> So, in your opinion, the current code base can handle out-of-order handling of packets?

I never said this.  I also agree it's technically necessary for your example.

Reply:
I know you never said this explicitly, you only said that it "is a stretch", and I was trying to guess what that means. Because this isn't a "stretchy" thing. Only true or false.

Reply:
all I intended to say was that it's not a technical shortcoming, we don't have/support asynchronous packet handling, shard goes brr either way

it's actually kind of weird that nobody has implemented it in however many years this pattern has been available in C# (assuming not the whole 20~ year life of this codebase)

it's a cool improvement, and it's simple, and I like it

Reply:
> all I intended to say was that it's not a technical shortcoming, we don't have/support asynchronous packet handling

But you replied to something that was *not* about asynchronous packet handling but about blocking packet handling to maintain ordering.
It looks like you mixed up this context, which is the root of our misunderstanding: I did not say async packet handling is necessary. I only said ordered/serial packet handling is technically necessary. And you fully agreed with this. (And really nothing is a stretch.)

Reply:
👍